Add IPv6 support

mrbesen opened the pull request:
This adds IPv6 support as mentioned in #39 .
It works on Linux with ICMP and HTTP-Ping. (Of cause IPv4 still works)

You can enter a host by its ipv4 address (`1.1.1.1`), its ipv6 address (`2a03:4000:2b:41f::4`) or by its hostname (`ipv4.mrbesen.de` or `ipv6.mrbesen.de`).
HTTP ping still works just by prepending `http://` and the Device option (`-I`) also works.

To allow for resolving ipv6 addresses i moved from `gethostbyname()` to `getaddrinfo()`. All the code for resolving the address is in `resolve.c`.

I could not get it to work on Windows. I tried different ways of compiling the code for windows. Crosscompiling it with Mingw created a exe file that kinda worked with wine but straight up crashed on my win10 vm (no matter the arguments). I implemented the old `gethostbyname()` resolve method for windows but it did not help. I have no experience developing C-Applications on Windows and i have no idea where to start the debugging, or even how or which debugger to use on Windows.

Maybe someone more "Windows-experienced" can take a look on the windows side of this PR.

I have not added any way of forcing ipv6 or ipv4 (for example a `-4` or `-6` option) the only way would be to just use the plain address instead of the hostname. I think that is fine for now and would be fairly easy to implement, because `getaddrinfo()` allows for a hint which can force the resolver to focus on ipv4 or ipv6.


P.S: 
Feel free to use my server as test target (`ipv4.mrbesen.de`, `ipv6.mrbesen.de` and `mrbesen.de`)

dreua commented on resolve.c:
```diff
@@ -0,0 +1,94 @@
+	{
+		// error - this should not happen
+		freeaddrinfo(res);
+		exit( -1 );
```
Maybe just add `ERRM("ai_addrlen too long\n");` or something like that.

mrbesen replied:
for that if beeing triggered `getaddrinfo()` would need to resolve the "service" to something other than `AF_INET (struct sockaddr_in)` or `AF_INET6 (struct sockaddr_in6)`. In the manpage of getaddrinfo i did not found any hint about that beeing possible.
But you are right, adding some kind of error message wont hurt.

dreua reviewed (commented):
Thank you, that looks really good and works nicely on my Fedora!

Do I understand correctly that this branch doesn't work at all on Windows? Could it be the changed include (ws2tcpip.h)?
I'm not a Windows expert either, unfortunately, but I think we should manage to get it working again...

mrbesen commented on ping.c:
```diff
@@ -271,21 +332,30 @@ void listener()
+		}
+		else // ipv6
+		{
+			offset = 8;
```
I dont know. I found the 8 with some trial and error.

mrbesen replied:
We could move the magic numbers away in a .h file (or some other place that no one cares. Top of the c file maybe?) like this:
```c
#ifdef __FreeBSD__
  #define ICMP4_PACKET_OFFSET (48)
  #define ICMP6_PACKET_OFFSET (8)
#else
  #define ICMP4_PACKET_OFFSET (28)
  #define ICMP6_PACKET_OFFSET (8)
#endif
```

than the code around the offset variable can be something like this:
```c
int offset = (addr.sin6_family == AF_INET) ? ICMP4_PACKET_OFFSET : ICMP6_PACKET_OFFSET;
```

mrbesen replied:
Maybe it would be even more readable to have some macros for the ipv6 and ipv4 checks.
Those could be used around the program.

```c
#define IS_IPv4( A ) ( ((struct sockaddr*) (a))->sa_family == AF_INET )
...
if ( IS_IPv4(&addr) ) {
   ....
}
```

mrbesen commented on resolve.c:
```diff
@@ -0,0 +1,94 @@
+
+	if( addr->sa_family != AF_INET)
+	{
+		ERRM( "Only ipv4 is supported on windows");
```
AFAIK `gethostbyname()` should only return IPv4 the if is more a sanity check, because i was not able to test this thorough.

mrbesen commented:
> Do I understand correctly that this branch doesn't work at all on Windows? Could it be the changed include (ws2tcpip.h)?

Yes, it does not work on Windows.
I just tried to add the `winsock2.h` everywhere where i include `ws2tcpip.h` but it did not change anything. 
It crashes instantly. No matter if i choose http, icmp, ipv4 or ipv6. Starting the program as admin does not help either.

How it looks to start cnping (the window closes itself shortly after):
![grafik](https://user-images.githubusercontent.com/8640309/190468415-3bdffdd5-8c32-47d9-83fc-a650a1440e25.png)

Without an argument it creates this screen:
![grafik](https://user-images.githubusercontent.com/8640309/190468801-4596504a-e714-4cfb-9975-39eaa3633e2d.png)

But after pressing "Ping" it goes back to the gray crash and closes 1s later.
There is never the error MessageBox you get from the `ERRMB` macro.

Maybe i just f-ed something up compiling the exe file.
(All though the file seems to work with wine - at least with ICMP Ping, HTTP Ping seems to be a bit broken too)
Is there a way to download the exe file from the Ci Pipeline?

dreua commented:
> Is there a way to download the exe file from the Ci Pipeline?

Currently no, I might convert that to GH Actions sooner or later, then yes. What happens iff you compile the current master, does that work?

mrbesen commented:
I just compiled the master and it did not work (same crash).
So looks like i am just not able to compile at all. (the 1.0.0 [release binary](https://github.com/cntools/cnping/releases/tag/1.0.0) works fine)
Do you have any tips on how to compile the binary?
I have made a simple docker container from the `ubuntu:bionic` image and then  installed the dependencys that the Ci installs too.
```
sudo apt update
sudo apt-get install -y make libxinerama-dev libxext-dev libx11-dev build-essential binutils-mingw-w64-i686 gcc-mingw-w64-i686 g++-mingw-w64-i686 mesa-common-dev libglvnd-dev
```
And then i just compiled cnping with `make cnping.exe`

Nothing fancy, i just dont want all the mingw stuff on my main system.

dreua commented:
I have no clue, your process sounds exactly right: As you said, that's what the CI does. I'll try to look into it and get back to you. Does an older version of cnping.exe work? I once had crashes because of missing drivers or such, especially on a VM.

mrbesen commented:
I have now removed the commit where i implemented the alternative (old) resolve method for windows and the commit where i reverted that change.

I would be happy if you could merge this.

dreua commented:
Hi sorry this has been on my todo list for some time now. I'd like to test it myself on Windows again but haven't found the time. I'm sure I can do that in the first half of December, would that be sufficient for you or do you need it merged faster for some reason?

mrbesen commented:
No, take your time.
